dnsmasq_interface.c

```diff
@@ -1415,7 +1377,44 @@ int FTL_listsfile(const char* filename, unsigned int index, FILE *f, int cache_s
 	gravityDB_finalizeTable();
 
 	// Final logging
-	logg("%s: imported %i domains (took %.1f ms)", filename, added, timer_elapsed_msec(LISTS_TIMER));
+	logg("Database (%s): imported %i domains (took %.1f ms)", tablename, added, timer_elapsed_msec(LISTS_TIMER));
+
+	return added;
+}
+
+int FTL_database_import(int cache_size, struct crec **rhash, int hashsz)
+{
+	struct all_addr addr4 = {{{ 0 }}}, addr6 = {{{ 0 }}};
+	bool has_IPv4 = false, has_IPv6 = false;
+
+	if(blockingstatus == BLOCKING_DISABLED)
+	{
+		logg("Skipping import of database tables because blocking is disabled");
+		return cache_size;
+	}
+
+	// Get IPv4/v6 addresses for blocking depending on user configured blocking mode
+	prepare_blocking_mode(&addr4, &addr6, &has_IPv4, &has_IPv6);
+
+	// If we have neither a valid IPv4 nor a valid IPv6 but the user asked for
+	// blocking modes MODE_IP or MODE_IP_NODATA_AAAA then we cannot add any entries here
+	if(!has_IPv4 && !has_IPv6)
+	{
+		logg("ERROR: Cannot add domains from gravity because pihole-FTL found\n" \
+		     "       neither a valid IPV4_ADDRESS nor IPV6_ADDRESS in setupVars.conf" \
+		     "       This is an impossible configuration. Please contact the Pi-hole" \
+		     "       support if you need assistance.");
+		return cache_size;
+	}
+
+	// Import gravity and blacklist domains
+	int added;
+	added  = FTL_table_import("gravity", GRAVITY_LIST, SRC_GRAVITY, addr4, addr6, has_IPv4, has_IPv6, cache_size, rhash, hashsz);
+	added += FTL_table_import("blacklist", BLACK_LIST, SRC_BLACK, addr4, addr6, has_IPv4, has_IPv6, cache_size, rhash, hashsz);
+
+	// Update counter of blocked domains
 	counters->gravity += added;
```

Comment:
Shouldn't this be an assignment, because we are re-importing gravity? Otherwise, every reload would inflate the size of the counter.

Reply:
We can make this an assignment without changing the effect. Before this function is run, we zero the counter here:

https://github.com/pi-hole/FTL/blob/076c4ef7d5f7de81388f6d79e2ab33477a5149ab/dnsmasq_interface.c#L372